### ISSUE TITLE
toggleLinkHintCharacters doesn't work on Wikipedia

### ISSUE THREAD

**Issue by mlololoer** [gdh1995/vimium-c#514]
**Describe the bug**

When viewing Wikipedia pages, toggleLinkHintCharacters fails to work properly every second page you visit.

**To Reproduce**

1. Reset Vimium C's settings, and then use the following custom key mappings:
```
unmapAll
map e toggleLinkHintCharacters value="abcdefg"
```
2. Go to any Wikipedia page, e.g. https://en.wikipedia.org/wiki/Semiosis (this bug happens on some other sites too, but it happens most reliably on Wikipedia)
3. Notice how pressing 'e' multiple times will allow you to toggle between the default link hint characters and "abcdefg" freely
4. Now click on any link on the same Wikipedia page
5. Now, pressing 'e' multiple times makes toggleLinkHintCharacters 'stuck' - you can't toggle between the default link hints and "abcdefg" freely anymore, the link hints will stay stuck on "abcdefg".

**Browser, OS and Vimium C versions**

- Browser name: Chrome
- Browser version: Version 96.0.4664.45 (Official Build) (64-bit)
- Vimium C version: 1.94.0
- OS name and version: Windows 10 10.0.19043 Build 19043

Also tested on Microsoft Edge on the same OS, and Chrome on Kali Linux, but this bug does NOT appear.

**Comment by gdh1995** [comment]
1. Does step 4 go to another URL, or just a section of the same page ?
2. Does `stuck` mean it always reports `Now linkHintCharacters uses "abcdefg"`? or just nothing happened ?
3. when you notice step 5, was an input area of the page focused ? If you add some other key mappings, will those keys work?

**Comment by mlololoer** [comment]
1. Yep, another URL on the same page, like below:
![step1](https://user-images.githubusercontent.com/58585450/144347324-d0f1576e-9f7b-4095-9ed8-90bcf5c88df8.png)

2. It reports `Now linkHintCharacters uses "abcdefg"` once (like below), but after that, pressing `e` doesn't do anything. The message doesn't appear again.
![step2](https://user-images.githubusercontent.com/58585450/144347328-ffd55414-eae7-4b99-8d35-a3cef530f0ae.png)

3. There wasn't any focused input area. Even if you add a different key (e.g. `s` for `scrollDown`), you can still scroll even though the linkHintCharacters is stuck.
```
unmapAll
map e toggleLinkHintCharacters value="abcdefg"
map s scrollDown
```

**Comment by gdh1995** [comment]
Sorry but I didn't reproduce it yet.

1. When it's "stuck", do commands like `LinkHints.activate*` work yet? Will they use `ABCDEFG` ?
2. What's your keyboard layout? English-QWERTY, French, German, Chinese, or anything else?
3. What's your real key mapping? Does your real mapping include <kbd>Alt</kbd>?
4. Please press `<F12>` to open Chrome's `Developer Tools` when it's "stuck", go to `Console` panel and put potential error logs here, like:

![image](https://user-images.githubusercontent.com/5547703/144367044-d99aacdd-faa5-463c-9cd8-c532e87c933b.png)

**Comment by mlololoer** [comment]
1. LinkHints.activate still works, but it will always use ABCDEFG, pressing `e` won't change the link hint characters
2. English (QWERTY).
3. The key mappings don't include any modifiers.
4. There aren't any abnormal error messages in the console log that I can see when I try playing with toggleLinkHintCharacters or LinkHints.activate.

**Comment by mlololoer** [comment]
I've been able to find a little more information on the bug - for some reason, the `onDisconnect` event of the page is being called some time before I activate `toggleLinkHintCharacters`.

(Consequently, `onDisconnect` goes through the `framesForTab_` array and deletes the current port from it, meaning that when I activate `toggleLinkHintCharacters`, it runs `kBgCmd.toggle` in `background/all_commands.ts`. In the last few lines of `kBgCmd.toggle`, the `framesForTab_` array is accessed to find the current port, but since the port was prematurely deleted, there is an error, which explains why `toggleLinkHintCharacters` was getting "stuck".)

I haven't been able to dig deeper into why the `onDisconnect` event is being fired prematurely, but I hope you could take one more look into this. Thanks

**Comment by gdh1995** [comment]
Thank you very much!



---Original---
From: ***@***.***&gt;
Date: Thu, Aug 4, 2022 00:20 AM
To: ***@***.***&gt;;
Cc: "Dahan ***@***.******@***.***&gt;;
Subject: Re: [gdh1995/vimium-c] toggleLinkHintCharacters&nbsp;doesn't work on Wikipedia (Issue #514)




 
I've been able to find a little more information on the bug - for some reason, the onDisconnect event of the page is being called some time before I activate toggleLinkHintCharacters.
 
(Consequently, onDisconnect goes through the framesForTab_ array and deletes the current port from it, meaning that when I activate toggleLinkHintCharacters, it runs kBgCmd.toggle in background/all_commands.ts. In the last few lines of kBgCmd.toggle, the framesForTab_ array is accessed to find the current port, but since the port was prematurely deleted, there is an error, which explains why toggleLinkHintCharacters was getting "stuck".)
 
I haven't been able to dig deeper into why the onDisconnect event is being fired prematurely, but I hope you could take one more look into this. Thanks

**Comment by gdh1995** [comment]
> since the current port was prematurely deleted, there is an error, which explains why `toggleLinkHintCharacters` was getting "stuck"

Sorry but is this line just your suppose? Did you once see any direct error messages related to the handler of `kBgCmd.toggle`?

In my design, `ports_` only includes alive ports, and I've reviewed `onDisconnect` in `background/ports.ts` but not found bugs.

**Comment by mlololoer** [comment]
Yes, there is an error message in `kBgCmd.toggle` (in `background/all_commands.ts`).
The issue isn't with `onDisconnect`, the issue is that the event for `onDisconnect` is getting fired early for some reason.

I’ll try to provide you with as much information as possible, since it seems like it is still not reproducible for you.
### Background
First I downloaded your repository, unzipped it, ran `npm install`, and then `gulp local` to compile a non-minified distribution to make debugging easier.

Then I zipped up the extension and loaded it into Chrome, and I changed Vimium C’s settings to use these key mappings:
```
unmapAll
map e toggleLinkHintCharacters value="abcdefg"
```
I’ll demonstrate the bug on Wikipedia (there are some other sites where this bug happens too, but it works most reliably on Wikipedia). Chrome version is 104.0.5112.79 (Official Build) (64-bit). OS is Windows 10. The only tab open is the Wikipedia page, and there are no other tabs nor windows that are opened in the browser.

Now, on the first page that I open on Wikipedia, everything will be ok. I can press e, which will toggle the link hints from “sadjklewcmpgh” to “abcdefg”. The HUD message will appear in the bottom right as well. Pressing e again will toggle them back to “sadjklewcmpgh”, without any problems.
![1](https://user-images.githubusercontent.com/58585450/183060207-4f3f498e-6087-48c7-a470-e90a4a3f9c5b.png)
![2](https://user-images.githubusercontent.com/58585450/183060247-5b21d30b-ec02-4e31-a591-1b6ebbedefdc.png)

### Bug
The bug appears when I go and click on another link within the page. For example, I will click “software development” near the top of the article.

Pressing e the first time will show the HUD in the bottom right as usual and switches the linkHints. 
![3](https://user-images.githubusercontent.com/58585450/183060300-a57da997-37c2-49d9-ac8d-6be65cf60ade.png)

But in the second time that I press e, the background.html of the extension shows an error in the `kBgCmd.toggle` function:
![4](https://user-images.githubusercontent.com/58585450/183060355-f838d0e5-2d60-4204-a257-c537378cd005.png)

I decided to find the difference between the first and second time I press e on the “software development” page. I looked into how `frames` is assigned, and I found that the `framesForTab_` map is empty on the second time that I press e, which is why there is an error reading `frames.cur_`.
On the first time I press e, the `store_1.framesForTab_` map contains an entry. (I assume it is the port object for the current tab). Since the entry exists, `frames.cur_` doesn’t cause an error.
![5](https://user-images.githubusercontent.com/58585450/183060420-90980a70-7ccb-4886-87ed-daf48708511b.png)

On the second time I press e, the `framesForTab_` map is now empty – the port has been deleted for some reason, and so the error occurs when reading `frames.cur_`.
![6](https://user-images.githubusercontent.com/58585450/183060502-0c35faf7-3408-4d24-a21e-0d3bb41d2eb9.png)

Subsequent error:
![7](https://user-images.githubusercontent.com/58585450/183060571-b7ac3d38-d8df-4703-8abd-ed0a20269062.png)

I tried to look into exactly why the port is disappearing from `framesForTab_` so I used data watchpoints to see where the `framesForTab_` object was being modified, and I found that it gets modified in the `onDisconnect` event handler in `background/ports.js`:
![8](https://user-images.githubusercontent.com/58585450/183060611-7505eea4-fe3f-467a-94d8-ed763a9b0a1d.png)

Line 128 seemed to be responsible for the deletion of the port object, so I put a breakpoint there to see when it was being triggered.

I tried reproducing the problem again, and after the first time I press e on the “software development” page, the onDisconnect event is fired and the extension breaks at line 128. So for some reason, the onDisconnect event is getting fired _even though we have not reloaded the page or clicked a different link – it is getting fired just because we toggled the linkHints._

I haven’t been able to debug what exactly is causing onDisconnect to fire, so this is as far as I’ve gotten.

If you need any more information, I will gladly provide it to you. If you could guide me on how to further investigate this problem, I would very much appreciate it.

**Comment by gdh1995** [comment]
Thanks for your patience. If `!port.s.frameId_` is true, then `.frameId_` should be `0`, which means the port is a "top"-level frame of a tab. And by default Vimium C's port should only be closed when a page is being closed, so the event of "a top-level frame is closed" should have been enough to indicate "the whole tab is disappearing / reloading".

Now I wonder:
* Do you have tested on a Chrome with all other extensions disabled? If you run `chrome.exe --user-data-dir=D:\UserData` then you can test it on a fresh profile of Chrome.
* What's your chrome and Vimium C versions now?

**Comment by gdh1995** [comment]
Oh I reproduced on a fresh profile of Chrome 104 and Vimium C v1.98.3...

I once mistakenly thought you clicked an same-page link and only make the page scrolls. Now I know "its content" will switch to a new page. Let me debug it.

**Comment by gdh1995** [comment]
OKay, it's prefetching causing a wrong order among different pairs of `connect` and `disconnect` events, and then `onConnect` runs into a wrong state.

I'll fix it in a next version of Vimium C.

Thank you again!

**Comment by gdh1995** [comment]
Hello, I've pushed a fix (commit f89f1b85) and then you may take a try.

**Comment by mlololoer** [comment]
Hi, I don't see the bug appearing anymore. Thanks for the help!
